runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = release()
+	}()
+	return d.createTableAsSelect(ctx, conn, release, name, query, opts)
+}
+
+func (d *db) createTableAsSelect(ctx context.Context, conn *sqlx.Conn, releaseConn func() error, name, query string, opts *CreateTableOptions) error {
+	// check if some older version exists
+	oldVersion, oldVersionExists, _ := tableVersion(d.writePath, name)
+	d.logger.Debug("old version", slog.String("version", oldVersion), slog.Bool("exists", oldVersionExists))
+
+	// create new version directory
+	newVersion := newVersion()
+	newVersionDir := filepath.Join(d.writePath, name, newVersion)
+	err := os.MkdirAll(newVersionDir, fs.ModePerm)
+	if err != nil {
+		return fmt.Errorf("create: unable to create dir %q: %w", name, err)
+	}
+
+	var m meta
+	if opts.View {
+		// create view - validates that SQL is correct
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE VIEW %s AS (%s\n)", safeSQLName(name), query))
+		if err != nil {
+			return err
+		}
+
+		m = meta{ViewSQL: query}
+	} else {
+		// create db file
+		dbFile := filepath.Join(newVersionDir, "data.db")
+		safeDBName := safeSQLName(dbName(name))
+
+		// detach existing db
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("DETACH DATABASE IF EXISTS %s", safeDBName), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: detach %q db failed: %w", safeDBName, err)
+		}
+
+		// attach new db
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("ATTACH %s AS %s", safeSQLString(dbFile), safeDBName), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: attach %q db failed: %w", dbFile, err)
+		}
+
+		// ingest data
+		_, err = conn.ExecContext(ctx, fmt.Sprintf("CREATE OR REPLACE TABLE %s.default AS (%s\n)", safeDBName, query), nil)
+		if err != nil {
+			_ = os.RemoveAll(newVersionDir)
+			return fmt.Errorf("create: create %q.default table failed: %w", safeDBName, err)
+		}
+
+		m = meta{Format: BackupFormatDB}
+	}
+
+	d.writeDirty = true
+	// write meta
+	err = writeMeta(newVersionDir, m)
+	if err != nil {
+		_ = os.RemoveAll(newVersionDir)
+		return err
+	}
+
+	// update version.txt
+	err = os.WriteFile(filepath.Join(d.writePath, name, "version.txt"), []byte(newVersion), fs.ModePerm)
+	if err != nil {
+		_ = os.RemoveAll(newVersionDir)
+		return fmt.Errorf("create: write version file failed: %w", err)
+	}
+
+	// close write handle before syncing read so that temp files or wal files if any are removed
+	err = releaseConn()
+	if err != nil {
+		return err
+	}
+
+	if err := d.syncBackup(ctx, name); err != nil {
+		return fmt.Errorf("create: replicate failed: %w", err)
+	}
+	d.logger.Debug("table created", slog.String("name", name))
+	// both backups and write are now in sync
+	d.writeDirty = false
+	if oldVersionExists {
+		_ = os.RemoveAll(filepath.Join(d.writePath, name, oldVersion))
+		_ = d.deleteBackup(ctx, name, oldVersion)
+	}
+	return d.syncRead(ctx)
```

Comment:
1. What if it exits early due to an error (like context cancelled) while `writeDirty` is true?
2. Why does it need to call the remove/delete functions? The word "sync" makes it sound like it should already take care of cleaning up files that no longer exist remotely?
3. Does calling `syncRead` from the write path put the read connections at risk of referencing removed files?

Reply:
1. In the next db update call, we will do a `syncWrite` which will update everything from the backup. There is a slight chance on inconsistency if `version.txt` is updated in backup and caller gets a `cancelled` error. I thought of updating `version.txt` with a non cancellable context but then avoided it because our API is a replace API.
2. `syncBackup` will just update the backup. The source of truth at this point is `write` so we need to delete this explicitly. We can also keep it as it is and the next `syncWrite` will ensure old version is deleted. Though I think `deleteBackup` should not be necessary. I think that is already handled in `syncBackup` and can be removed from there. I will reconfirm.
3. No. If you look at the `attachDBs` we always look at the write directory to attach dbs to read handles. The tables from read directory are only removed when new read handle is created and only one destroyed.
```
// NOTE :: we always look at the write version
// Tables in read path are removed after getting a new handle
// So we need to always look at the write version to ensure we do not reattach dropped tables
```

Reply:
1. I think the sequence needs to be: create local table, upload table to bucket, update version.txt in bucket, update version.txt on local. Basically, it shouldn't serve reads from the new version until after it has been fully committed to the bucket. Otherwise, it opens up for inconsistencies where a cancellation or restart at the wrong time can lead it to some queries being served from a new table, and then it reverts back to the old table, right?
2. Sounds good. I think the bucket should be considered the source of truth and the local copy should be considered a cache. Essentially we only want to serve reads that we know are fully committed to the bucket. In terms of [Postgres transaction isolation](https://www.postgresql.org/docs/current/transaction-iso.html), this would be a "read committed" guarantee.